src/data/matrix/reindex.lean

```diff
@@ -0,0 +1,40 @@
+/-
+Copyright (c) 2021 Filippo Nuccio. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Filippo Nuccio
+-/
+import data.matrix.basic
+import linear_algebra.matrix.reindex
+
+/-!
+# Matrices
+-/
+universes v
+
+open matrix
+
+variables {m n o : Type*} [fintype m] [fintype n] [fintype o]
+variables {m₁ n₁ o₁ : Type*} [fintype m₁] [fintype n₁] [fintype o₁]
+variables {R : Type*}{α : Type v}
+
+namespace matrix
+
+/-- equivalences between matrix ((m × n) × o) ((m₁ × n₁) × o₁) R and
+ matrix (m × n × o) (m₁ × n₁ × o₁) R -/
+def index_assoc : matrix ((m × n) × o) ((m₁ × n₁) × o₁) R ≃ matrix (m × n × o) (m₁ × n₁ × o₁) R :=
+matrix.reindex (equiv.prod_assoc _ _ _) (equiv.prod_assoc _ _ _)
+
+def linear_equiv_index_assoc [comm_semiring α] [add_comm_monoid R] [module α R] :
+  matrix ((m × n) × o) ((m₁ × n₁) × o₁) R ≃ₗ[α] matrix (m × n × o) (m₁ × n₁ × o₁) R :=
```

Comment:
The reason you get the error is because `reindex_linear_equiv` produces the stronger `R`-linear equiv; there is no need to introduce `α`.

The caller can always weaken the strength of the linearity using `linear_equiv.restrict_scalars`.

Reply:
```suggestion
def linear_equiv_index_assoc :
  matrix ((m × n) × o) ((m₁ × n₁) × o₁) R ≃ₗ[R] matrix (m × n × o) (m₁ × n₁ × o₁) R :=
```

Reply:
Oh, I now see that your statement is actually more general - I think you should modify `matrix.reindex_linear_equiv` to use your formulation with an extra (explicit) `α` variable!

Reply:
Done in #8159